templating/base_render_options.go

```diff
@@ -152,6 +159,54 @@ func LoadAndRenderSteps(template *Template, opts *BaseRenderOptions) (string, er
 	return rendered, nil
 }
 
+// ResolveAndMergeSecrets parses the secrets in the template, resolves them using vault providers and adds them to the values map.
+func resolveAndMergeSecrets(ctx context.Context, template *Template, secretResolver *SecretResolver, valueMap *Values) error {
+
+	// Cheap optimization to skip the secrets merging if it doesn't contain "secrets" string in it. Note that the task can
+	// have the string secrets but may not essentially the secrets section.
+	if !strings.Contains(string(template.Data), "secrets") {
+		return nil
+	}
+
+	// At first render the template with existing values to render templatized vaules for secrets.
+	values := *valueMap
+	values["Secrets"] = Values{}
+	engine := NewEngine()
```

Comment:
nit: can save an allocation on the engine by re-using it between the double render